### ISSUE TITLE
[restrict-plus-operands] False positive when using isNaN to check type of argument

### ISSUE THREAD

**Issue by hrb-bausmert-robert** [typescript-eslint/typescript-eslint#630]
**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/restrict-plus-operands": "error",
  }
}
```

```TS
function restrictPlusOperandsTest(numberArgument?: number) {
  if (!isNaN(numberArgument)) {
    numberArgument + 1
  }
}

function restrictPlusOperandsTest2(numberArgument: number | null) {
  if (!isNaN(numberArgument)) {
    numberArgument + 1
  }
}

function restrictPlusOperandsTest3(numberArgument: number | string) {
  if (!isNaN(numberArgument)) {
    numberArgument + 1
  }
}
```

**Expected Result**
I expect the rule to understand that the if clause will only be executed if the passed argument is a number

**Actual Result**
Error: Operands of '+' operation must either be both strings or both numbers.eslint(@typescript-eslint/restrict-plus-operands)

**Additional Info**
This error seems to occur whenever the argument passed into a function is either a union type or an optional argument, one of the possible argument types is number and an if check with isNaN is performed on the argument.

Explicitly checking for the non-number type the argument can be does not result in an error. The typescript-eslint parser does not error for the following functions:
```
function restrictPlusOperandsTest(numberArgument?: number) {
  if (numberArgument !== undefined) {
    numberArgument + 1
  }
}

function restrictPlusOperandsTest2(numberArgument: number | null) {
  if (numberArgument !== null) {
    numberArgument + 1
  }
}

function restrictPlusOperandsTest3(numberArgument: number | string) {
  if (typeof numberArgument !== 'string') {
    numberArgument + 1
  }
}
```

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `1.9.0` |
| `@typescript-eslint/parser`        | `1.9.0` |
| `TypeScript`                       | `3.3.3333` |
| `ESLint`                           | `5.16.0` |
| `node`                             | `10.14.1` |
| `npm`                              | `6.4.1` |

**Comment by bradzacher** [comment]
`isNaN` is not typed with a type guard in the typescript standard library

https://github.com/microsoft/TypeScript/blob/9c71eaf59040ae75343da8cdff01344020f5bba2/lib/lib.es2015.core.d.ts#L235-L241

This means that typescript ***does not narrow the type when using isNaN***

If you are using `strictNullChecks` this all becomes a lot clearer

```ts
function restrictPlusOperandsTest(numberArgument?: number) {
  if (!isNaN(numberArgument)) {
    // typeof numberArgument === number | undefined
    numberArgument + 1 // error: (number | undefined) + number
  }
}

function restrictPlusOperandsTest2(numberArgument: number | null) {
  if (!isNaN(numberArgument)) {
    // typeof numberArgument === number | null
    numberArgument + 1 // error: (number | null) + number
  }
}

function restrictPlusOperandsTest3(numberArgument: number | string) {
  if (!isNaN(numberArgument)) { // this doesn't even work in typescript because as per the definition above, isNaN only accepts number
    // typeof numberArgument === number | string
    numberArgument + 1 // error: (number | string) + number
  }
}
```


Just open the [example in the typescript playground](https://www.typescriptlang.org/play/#src=function%20restrictPlusOperandsTest(numberArgument%3F%3A%20number)%20%7B%0A%20%20if%20(!isNaN(numberArgument))%20%7B%0A%20%20%20%20numberArgument%20%2B%201%0A%20%20%7D%0A%7D%0A%0Afunction%20restrictPlusOperandsTest2(numberArgument%3A%20number%20%7C%20null)%20%7B%0A%20%20if%20(!isNaN(numberArgument))%20%7B%0A%20%20%20%20numberArgument%20%2B%201%0A%20%20%7D%0A%7D%0A%0Afunction%20restrictPlusOperandsTest3(numberArgument%3A%20number%20%7C%20string)%20%7B%0A%20%20if%20(!isNaN(numberArgument))%20%7B%0A%20%20%20%20numberArgument%20%2B%201%0A%20%20%7D%0A%7D%0A), and turn on the `strictNullChecks` argument from the options dropdown and you'll see that the code you've provided is very invalid and all the types are wrong.
![image](https://user-images.githubusercontent.com/7462525/59937718-14230780-9408-11e9-8cc7-e8c98cf405a2.png)
